Add detection of unknown fields for policies (CNP & CCNP) in preflight

christarazi opened the pull request:
~Please only pay attention to the three most recent commits. This PR depends on
https://github.com/cilium/cilium/pull/11477 and includes commits from it (until it's merged.)~ See commit msgs.

Example output of "unknown fields found":
```
$ kubectl get cnp cnp-update -o yaml
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
...
  name: cnp-update
...
spec:
  endpointSelector: {}
  wrong: foo

$ kubectl -n cilium exec -it cilium-ffdz4 -- bash
root@gke-chris-cluster-default-pool-d04d0083-8xx4:/home/cilium# cilium preflight validate-cnp
INFO[0000] Setting up Kubernetes client
INFO[0000] Establishing connection to apiserver          host="https://10.47.240.1:443" subsys=k8s
INFO[0000] Connected to apiserver                        subsys=k8s
INFO[0000] Unable to retrieve EndpointSlices for default/kubernetes. Disabling EndpointSlices  error="the server could not find the requested resource (get endpointslices.discovery.k8s.io kubernetes)" subsys=k8s
WARN[0000] It seems you have a policy with extra unknown fields. Consider removing these fields, as they have no effect. The presence of these fields may have introduced a false sense security, so please check whether your policy is actually behaving as you expect.  ciliumNetworkPolicyName=cnp-update subsys=validator
ERRO[0000] Unexpected validation error                   CiliumNetworkPolicy=default/cnp-update error="unknown fields found, extra:[spec.wrong]"
ERRO[0000] Found invalid CiliumNetworkPolicy
```

Example output of "top-level description field found":
```
$ kubectl get cnp cnp-update -o yaml
apiVersion: cilium.io/v2
description: Policy with the same name as cnp-update-deny-ingress to test CNP update
kind: CiliumNetworkPolicy
metadata:
...
  name: cnp-update
...
spec:
  endpointSelector: {}

$ kubectl -n cilium exec -it cilium-ffdz4 -- bash
root@gke-chris-cluster-default-pool-d04d0083-8xx4:/home/cilium# cilium preflight validate-cnp
INFO[0000] Setting up Kubernetes client
INFO[0000] Establishing connection to apiserver          host="https://10.47.240.1:443" subsys=k8s
INFO[0000] Connected to apiserver                        subsys=k8s
INFO[0000] Unable to retrieve EndpointSlices for default/kubernetes. Disabling EndpointSlices  error="the server could not find the requested resource (get endpointslices.discovery.k8s.io kubernetes)" subsys=k8s
WARN[0000] It seems you have a policy with a top-level description. This field is no longer supported. Please migrate your policy's description field under `spec` or `specs`.  ciliumNetworkPolicyName=cnp-update subsys=validator
ERRO[0000] Unexpected validation error                   CiliumNetworkPolicy=default/cnp-update error="top-level description field found"
ERRO[0000] Found invalid CiliumNetworkPolicy
```

Updates: https://github.com/cilium/cilium/issues/12737

christarazi committed:
preflight: Refactor CCNP wildcard check

This is a non-functional change to move the validation check to a
separate function. This is a preparatory commit for future validation
checks to be added.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
preflight: Add unknown fields check for validation

This commit implements two additional validation checks for
`validate-cnp` preflight command:
  1) Presence of a top-level description field.
  2) Presence of extra unknown fields.

This commit relies on the fork
github.com/christarazi/structured-merge-diff, to workaround a panic when
converting a CCNP into an `unstructured.Unstructured` type. See
https://github.com/kubernetes-sigs/structured-merge-diff/issues/172 for
details.

Fixes: https://github.com/cilium/cilium/issues/13155

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
vendor: Add github.com/jeremywohl/flatten v1.0.1

This is needed for an upcoming commit to validate that policies do not
contain unknown fields.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/k8s/apis/cilium.io/v2/validator/unknown_fields.go:
```diff
@@ -0,0 +1,236 @@
+
+func igIgnoredField(k string) bool {
+	return k == "metadata.creationTimestamp" || strings.HasPrefix(k, "spec.endpointSelector.matchLabels")
+}
```
Maybe add explanation why these are ignored?

sayboras replied:
and should it be `isIgnoredField` instead of `igIgnoredField` ?

christarazi replied:
:man_facepalming: How did I not notice that. :sweat_smile:

aanm commented on Documentation/concepts/kubernetes/configuration.rst:
```diff
@@ -258,10 +258,10 @@ will return an error, e.g.:
 	cat <<EOF > ./bad-cnp.yaml
```
There are some "`<redacted>`" in the commit message.

christarazi replied:
As per the PR description, this is leftovers from depending on another PR. These commits will be removed when I rebase.

aanm commented on pkg/k8s/apis/cilium.io/v2/validator/unknown_fields.go:
```diff
@@ -0,0 +1,236 @@
+		return fmt.Errorf("unknown kind %q", kind)
+	}
+
+	if _, ok := policy.Object["description"]; ok {
+		scopedLog.Warn(warnTopLevelDescriptionField)
+		return ErrTopLevelDescriptionFound
+	}
```
We don't need to specifically check for the `"description"` field. A [C]CNP can only have these top level fields: `apiVersion`, `kind`, `metadata` and `spec` or `specs`. If any key of the `policy.Object` map contains a key that does not belong to any of these fields then we should return an error.

christarazi replied:
Discussed offline. We are keeping this check because we want to have a specialized warning message about top-level description, so that we can advise users to modify their policies, instead of removing the "unknown" fields, which is what we recommend for any other field found.

aanm commented on pkg/k8s/apis/cilium.io/v2/validator/unknown_fields.go:
```diff
@@ -0,0 +1,236 @@
+		}
+	default:
+		// We've already validated above that there can only be two kinds: CNP &
+		// CCNP.
```
We can panic here, this will be a developer mistake.

aanm commented on pkg/k8s/apis/cilium.io/v2/validator/unknown_fields.go:
```diff
@@ -0,0 +1,236 @@
+}
+
+func igIgnoredField(k string) bool {
+	return k == "metadata.creationTimestamp" || strings.HasPrefix(k, "spec.endpointSelector.matchLabels")
```
Why are we ignoring these fields? Please add a comment describing it.

aanm commented on pkg/k8s/apis/cilium.io/v2/validator/unknown_fields.go:
```diff
@@ -0,0 +1,246 @@
+	default:
+		// We've already validated above that there can only be two kinds: CNP
+		// & CCNP. This is likely to be a developer error if hit, so panic.
+		scopedLog.WithField("kind", kind).Panic("Unexpected kind found when processing policy")
```
sorry I meant to say fatal `s/Panic/Fatal`

aanm commented on pkg/k8s/apis/cilium.io/v2/validator/unknown_fields.go:
```diff
@@ -0,0 +1,246 @@
+	// "spec.endpointSelector.matchLabels.key:app" to
+	// "spec.endpointSelector.matchLabels.any:key:app". This change is also
+	// insignificant and should not be warned about.
+	return f == "metadata.creationTimestamp" || strings.HasPrefix(f, "spec.endpointSelector.matchLabels")
```
thanks for the comment. What about the `spec.nodeSelector.matchLabels`? Also, the `spec.*Selector.requirements` might suffer from the same marshaling / unmarshaling logic.

In case someone adds another field with that uses the `matchLabels` and forgets to add it to this "ignored field" function, do we have a unit test covering this case?

christarazi replied:
Ah, I forgot about `nodeSelector`. Will add that and the `requirements`. For the unit test, if someone adds a field under `matchLabels`, it will also be ignored because we check it against the prefix.

christarazi replied:
Discussed offline. The summary is that there can be many field paths that have `matchLabels` in (include `matchExpressions`) and also arbitrary number of labels within it as well. Fixed with a regex solution.

nebril commented on pkg/k8s/apis/cilium.io/v2/validator/unknown_fields.go:
```diff
@@ -0,0 +1,246 @@
+	return fmt.Sprintf("unknown fields found, extra:%v", e.extras)
+}
+
+func toUnstructured(obj interface{}, clusterwide bool) (*unstructured.Unstructured, error) {
```
I know you already wrote this, but if you decide to use something ready-made, you might want to use `"k8s.io/apimachinery/pkg/runtime.DefaultUnstructuredConverter.ToUnstructured"`.

aanm replied:
ah! maybe this will solve the problem with marshaling / unmarshaling?

christarazi replied:
Thanks! I wasn't aware of this library. It doesn't solve our marshalling / unmarshalling problem, but I did use it to avoid duplicating the `toStructured` functionality. I did end up finding a bug in the library which I've fixed in a fork. Not sure if it's worth keeping, but I've vendored the fork in case.

nebril reviewed (approved):
Looks good overall. I think that using the library from `k8s.io/apimachinery/pkg/runtime` for conversion to `Unstructured` would be cleaner, but that's not that big of a deal.

fristonio reviewed (approved):
LGTM, apart from already requested changes. :rocket:

christarazi committed:
vendor: Use fork of structured-merge-diff

The intention of using this library is to utilize the facility to
convert Kubernetes objects to `unstructured.Unstructured`. Previously,
we were doing this by hand. For an explanation for why we are using a
fork, read below.

This commit migrates our use of sigs.k8s.io/structured-merge-diff/v4
over to the fork at github.com/christarazi/structured-merge-diff/v4
because of
https://github.com/kubernetes-sigs/structured-merge-diff/issues/172.
The fork contains the fix
(https://github.com/kubernetes-sigs/structured-merge-diff/pull/173)
which we are vendoring in.

Once the issue is fixed or the upstream PR is merged, we can revert back
to using the original library.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
